exp/topology/desiredstate/lifecycle_hooks.go

```diff
@@ -140,6 +180,97 @@ func (g *generator) callAfterControlPlaneUpgradeHook(ctx context.Context, s *sco
 	return true, nil
 }
 
+// callBeforeWorkersUpgradeHook calls the BeforeWorkersUpgrade before workers starts picking up a new worker version,
+// no matter if this is an intermediate versions of an upgrade plan or the target version of an upgrade plan.
+// NOTE: computeControlPlaneVersion records intent to call this hook when picking up a new control plane version
+// that exists also in the workers upgrade plan.
+func (g *generator) callBeforeWorkersUpgradeHook(ctx context.Context, s *scope.Scope, currentVersion *string, nextVersion string) (bool, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// NOTE: the hook should always be called before piking up a new version.
```

Comment:
Let's move this into the if directly above hookRequest. We should only run this code if we want to call the hook